lib/internal/webstreams/util.js

```diff
@@ -217,6 +220,54 @@ function lazyTransfer() {
   return transfer;
 }
 
+function createAsyncFromSyncIterator(syncIteratorRecord) {
+  const syncIterable = {
+    [SymbolIterator]: () => syncIteratorRecord.iterator,
+  };
+
+  const asyncIterator = (async function* () {
+    return yield* syncIterable;
+  }());
+
+  const nextMethod = asyncIterator.next;
+  return { iterator: asyncIterator, nextMethod, done: false };
+}
+
+function getIterator(obj, kind = 'sync', method) {
+  if (method === undefined) {
+    if (kind === 'async') {
+      method = obj[SymbolAsyncIterator];
+      if (method === undefined) {
+        const syncMethod = obj[SymbolIterator];
+        const syncIteratorRecord = getIterator(obj, 'sync', syncMethod);
+        return createAsyncFromSyncIterator(syncIteratorRecord);
+      }
+    } else {
+      method = obj[SymbolIterator];
+    }
+  }
+
+  const iterator = FunctionPrototypeCall(method, obj);
+  if (typeof iterator !== 'object' || iterator === null) {
+    throw new ERR_INVALID_STATE.TypeError('The iterator method must return an object');
+  }
+  const nextMethod = iterator.next;
+  return { iterator, nextMethod, done: false };
+}
```

Comment:
I wonder if we should use the V8-defined method instead (`%CreateAsyncFromSyncIterator`, and `getIterator` is not exposed directly but is still defined in C++), to avoid code duplication and reduce maintenance burden

https://github.com/nodejs/node/blob/f226350fcbebd4449fb0034fdaffa147e4de28ea/deps/v8/src/codegen/code-stub-assembler.h#L3333-L3341

Reply:
So can these methods be exposed to JS? i think would even make performance better

Reply:
Well yes you can use `%CreateAsyncFromSyncIterator` when you pass the `--allow-natives-syntax` flag – we'd need a way to use it without the flag, I haven't looked it up but that should be feasible, if not directly in the `.js` flle using `internalBinding`.

Reply:
Hmmmm, i looked at the v8 part i doesn't seem to be exposed to embedders, I assume that inclusions in "v8.h" are available to us?

Reply:
So I asked Joyee about that and she told me that the V8 team has been very clear those are internal only and us embedders cannot depend on it – except for testing purposes, if we pass the `--allow-natives-syntax` flag.

Reply:
Ah I see! marking this resolved then? 😄